ctru-rs/src/error.rs

```diff
@@ -1,11 +1,46 @@
 use std::error;
 use std::ffi::CStr;
 use std::fmt;
+use std::ops::{ControlFlow, FromResidual, Try};
 
 use ctru_sys::result::{R_DESCRIPTION, R_LEVEL, R_MODULE, R_SUMMARY};
 
 pub type Result<T> = ::std::result::Result<T, Error>;
 
+#[derive(Debug, Clone, PartialEq, PartialOrd)]
+#[repr(transparent)]
+pub(crate) struct LibCtruResult(pub i32);
+
+impl Try for LibCtruResult {
+    type Output = ();
+    type Residual = crate::Result<core::convert::Infallible>;
+
+    fn from_output(_: Self::Output) -> Self {
+        Self(0)
+    }
+
+    fn branch(self) -> ControlFlow<Self::Residual, Self::Output> {
+        if self.0 < 0 {
+            ControlFlow::Break(Err(self.into()))
+        } else {
+            ControlFlow::Continue(())
+        }
+    }
+}
+
+impl FromResidual for LibCtruResult {
+    fn from_residual(e: <Self as Try>::Residual) -> Self {
+        if let Some(e) = e.err() {
+            match e {
+                Error::Os(result) => Self(result),
+                _ => Self(-1),
+            }
+        } else {
+            Self(-1)
```

Comment:
This is a worst case scenario, right? In no normal case it should be an `Ok`, yet this is what it's implying. Maybe an `unreachable!` would be better? I didn't check if there are actual cases under which it is possible to run this line.

Reply:
Yeah, you're right, this shouldn't ever happen, I'll unwrap the error directly instead